[useNumberInput][base-ui] Fix change handlers passed through slotProps (#39407)
mui/material-ui: docs/data/base/components/number-input/number-input.md

/>
```

> ### Events

Reviewer: Does this only apply to the component (so, not also the hook)? If so then this is a good place for this content; if it also applies to the hook, then it should go under the "Customization" header.

Author: Hmm, a small part of this also applies to the hook but that part is quite important so I think it may be better to duplicate it

Reviewer: In that case, I think I would prefer to follow the pattern we've already established and just have one Events section under Customization that applies to both components and hooks. The Hook section has the text about how hooks can't use `slots` or `slotProps`, but we could clarify that again in the Events section if you think it's needed. So then text like "`onChange` can only be passed as a prop on the component—not through `slotProps`" could become "When working with the component, `onChange` can only be passed..."

Author: 👌 ~ Updated and moved to Customization: https://deploy-preview-39407--material-ui.netlify.app/base-ui/react-number-input/#events

> The NumberInput component provides two props - `onChange` and `onInputChange` - that accept event handlers for when the value of the component changes.

Reviewer: ```suggestion
The Number Input component provides two props—`onChange` and `onInputChange`—that accept event handlers for when the value of the component changes.
```

Author: Ah I always forget this one, wonder if `markdownlint` can detect it

Reviewer: ooh yeah it would be nice to have a rule to enforce it!

Reviewer: You can't really see the change in monospaced GH text, but be sure to change these en-dashes ("-") to em-dashes ("—") in addition to removing the spaces.

#### onChange

> `onChange` accepts a custom event handler that is called with two arguments - the underlying event, and the latest "clamped" value:

Reviewer: ```suggestion
`onChange` accepts a custom event handler that is called with two arguments: the underlying event, and the latest "clamped" value.
```

```ts
onChange: (
  event: React.FocusEvent<HTMLInputElement> | React.PointerEvent | React.KeyboardEvent,
  value: number | undefined,
) => void;
```

> It's called when the `<input>` element is blurred, or when the stepper buttons are clicked, after the value has been clamped based on the min, max, or step props.

Reviewer: ```suggestion
It's called when the `<input>` element is blurred, or when the stepper buttons are clicked, after the value has been clamped based on the `min`, `max`, or `step` props.
```

Reviewer: we might want to link those prop names to the corresponding sections since this would be the first mention of them on this page if it stays up here under `Component`

> Be aware that `onChange` can only be passed as a prop on the component, and not through `slotProps`.

Reviewer: ```suggestion
`onChange` can only be passed as a prop on the component—not through `slotProps`.
```

Reviewer: maybe this should go in a `:::warning` callout since it's devs need to "be aware" to avoid problems

```jsx
// ✅ Works
<NumberInput
  onChange={(event, newValue) => console.log(`${event.type} event: the new value is ${newValue}`)}
/>

// ❌ Doesn't work
<NumberInput
  slotProps={{
    input: {
      // expects a native input change event handler, newValue is always undefined
      onChange: (event, newValue) => { ... },
    },
  }}
/>
```

#### onInputChange

> `onInputChange` accepts a native input change handler that is passed to the `<input>` element:

Reviewer: ```suggestion
`onInputChange` accepts a native input change handler that's passed to the `<input>` element:
```

```ts
onInputChange: React.ChangeEventHandler<HTMLInputElement>;
```

> It's called whenever the value of the textbox changes, e.g. on every keystroke typed into it, before clamping is applied.

Reviewer: ```suggestion
It's called whenever the value of the textbox changes—for example, on every keystroke typed into it, before clamping is applied.
```

> In other words, it is possible for `event.target.value` to contain out-of-range values or non-numerical characters.

Reviewer: ```suggestion
In other words, it's possible for `event.target.value` to contain out-of-range values or non-numerical characters.
```

> Be aware that `onInputChange` can only be passed as a prop on the component. If you prefer to use slotProps, pass it as `slotProps.input.onChange` instead.

Reviewer: ```suggestion
`onInputChange` can only be passed as a prop on the component. 
If you prefer to use `slotProps`, pass it as `slotProps.input.onChange` instead.
```

Reviewer: same as above: `:::warning`?

  value: number | undefined,
) => void;
```

It's called when the `<input>` element is blurred, or when the stepper buttons are clicked, after the value has been clamped based on the [`min`, `max`](#minimum-and-maximum), or [`step`](#incremental-steps) props.

:::warning
When using the component, `onChange` can only be passed as a prop on the component—not through `slotProps`.
:::

```jsx
// ✅ Works
<NumberInput
  onChange={(event, newValue) => console.log(`${event.type} event: the new value is ${newValue}`)}
/>

// ❌ Doesn't work
<NumberInput
  slotProps={{
    input: {
      // expects a native input change event handler, newValue is always undefined
      onChange: (event, newValue) => { ... },
    },
  }}
/>
```

#### onInputChange

`onInputChange` accepts a native input change handler that's passed to the `<input>` element:

```ts
onInputChange: React.ChangeEventHandler<HTMLInputElement>;
```

It's called whenever the value of the textbox changes–for example, on every keystroke typed into it, before clamping is applied.

In other words, it's possible for `event.target.value` to contain out-of-range values or non-numerical characters.

> :::warning

Reviewer 2: IMO it could be an "info".

I'd also underline that both ways (`onInputChange` and `slotProps.input.onChange`) work the same way.

Author: Updated!
